### ISSUE TITLE
Crashes in  `NetworkInstrumentationFeature.extractTrace` and ` NetworkInstrumentationFeature.intercept`

### ISSUE THREAD

**Issue by JaviSoto** [DataDog/dd-sdk-ios#1638]
### The crash

We recently set up distributed network tracing, and we're hitting these crashes in production, although we're not able to reproduce, so they don't happen 100% of the time. Here's a few different stack traces, it's not always the same crash, but they happen in the same codepath:

<details><summary>Stack traces</summary>
<p>

```
OS Version: iOS 17.2 (21C62)
Report Version: 104

Exception Type: EXC_BREAKPOINT (SIGTRAP)
Crashed Thread: 10

Application Specific Information:
allHTTPHeaderFields

Thread 10 Crashed:
0   CoreFoundation                  0x31fe4ac54         CF_IS_OBJC
1   CoreFoundation                  0x31fe694e4         [inlined] CFDictionaryGetCount
2   CoreFoundation                  0x31fe694e4         CFDictionaryGetCount
3   CFNetwork                       0x3221fb2b8         _CFHTTPServerResponseEnqueue
4   CFNetwork                       0x3221e2be0         CFHTTPCookieStorageUnscheduleFromRunLoop
5   Foundation                      0x31e1a2718         URLRequest.allHTTPHeaderFields.getter
6   <AppName>                         0x104ec81c4         NetworkInstrumentationFeature.extractTrace (NetworkInstrumentationFeature.swift:246)
7   <AppName>                         0x104ec6ba4         [inlined] NetworkInstrumentationFeature.extractTrace
8   <AppName>                         0x104ec6ba4         NetworkInstrumentationFeature.intercept (NetworkInstrumentationFeature.swift:169)
9   <AppName>                         0x104ec6dbc         thunk for closure
10  libdispatch.dylib               0x32fe246a4         _dispatch_call_block_and_release
11  libdispatch.dylib               0x32fe262fc         _dispatch_client_callout
12  libdispatch.dylib               0x32fe2d890         _dispatch_lane_serial_drain
13  libdispatch.dylib               0x32fe2e3c0         _dispatch_lane_invoke
14  libdispatch.dylib               0x32fe39000         _dispatch_root_queue_drain_deferred_wlh
15  libdispatch.dylib               0x32fe38874         _dispatch_workloop_worker_thread
16  libsystem_pthread.dylib         0x3f3c2d960         _pthread_wqthread
```

```
OS Version: iOS 17.2.1 (21C66)
Report Version: 104

Exception Type: EXC_BREAKPOINT (SIGTRAP)
Crashed Thread: 7

Application Specific Information:
Exception 6, Code 1, Subcode 6598839380

Thread 7 Crashed:
0   CoreFoundation                  0x311296c54         CF_IS_OBJC
1   CoreFoundation                  0x3112a22fc         [inlined] CFRetain
2   CoreFoundation                  0x3112a22fc         CFRetain
3   CFNetwork                       0x31351c168         _CFStreamErrorFromCFError
4   CFNetwork                       0x31351c630         _CFStreamErrorFromCFError
5   CFNetwork                       0x31364f8d0         _CFHTTPServerResponseEnqueue
6   CFNetwork                       0x3134fe954         CFURLConnectionInvalidateConnectionCache
7   CFNetwork                       0x31362bef4         CFHTTPCookieStorageUnscheduleFromRunLoop
8   Foundation                      0x30f305000         URLRequest._unconditionallyBridgeFromObjectiveC
9   <AppName>                         0x10763a7f8         NetworkInstrumentationFeature.intercept (NetworkInstrumentationFeature.swift:155)
10  <AppName>                         0x10763adbc         thunk for closure
11  libdispatch.dylib               0x3212786a4         _dispatch_call_block_and_release
12  libdispatch.dylib               0x32127a2fc         _dispatch_client_callout
13  libdispatch.dylib               0x321281890         _dispatch_lane_serial_drain
14  libdispatch.dylib               0x3212823c0         _dispatch_lane_invoke
15  libdispatch.dylib               0x32128d000         _dispatch_root_queue_drain_deferred_wlh
16  libdispatch.dylib               0x32128c874         _dispatch_workloop_worker_thread
17  libsystem_pthread.dylib         0x3e52d1960         _pthread_wqthread
```

```
OS Version: iOS 17.2.1 (21C66)
Report Version: 104

Exception Type: EXC_BAD_ACCESS (SIGSEGV)
Exception Codes: SEGV_NOOP at 0x0000000000000120
Crashed Thread: 10

Application Specific Information:
allHTTPHeaderFields > count >
KERN_INVALID_ADDRESS at 0x120.

Thread 10 Crashed:
0   libobjc.A.dylib                 0x3035cc820         objc_msgSend
1   CFNetwork                       0x3152d32b8         _CFHTTPServerResponseEnqueue
2   CFNetwork                       0x3152babe0         CFHTTPCookieStorageUnscheduleFromRunLoop
3   Foundation                      0x31127a718         URLRequest.allHTTPHeaderFields.getter
4   <AppName>                         0x1033c81c4         NetworkInstrumentationFeature.extractTrace (NetworkInstrumentationFeature.swift:246)
5   <AppName>                         0x1033c6ba4         [inlined] NetworkInstrumentationFeature.extractTrace
6   <AppName>                         0x1033c6ba4         NetworkInstrumentationFeature.intercept (NetworkInstrumentationFeature.swift:169)
7   <AppName>                         0x1033c6dbc         thunk for closure
8   libdispatch.dylib               0x322f046a4         _dispatch_call_block_and_release
9   libdispatch.dylib               0x322f062fc         _dispatch_client_callout
10  libdispatch.dylib               0x322f0d890         _dispatch_lane_serial_drain
11  libdispatch.dylib               0x322f0e3c0         _dispatch_lane_invoke
12  libdispatch.dylib               0x322f19000         _dispatch_root_queue_drain_deferred_wlh
13  libdispatch.dylib               0x322f18874         _dispatch_workloop_worker_thread
14  libsystem_pthread.dylib         0x3e6f65960         _pthread_wqthread
```

```
OS Version: iOS 17.2.1 (21C66)
Report Version: 104

Exception Type: EXC_BAD_ACCESS (SIGSEGV)
Exception Codes: SEGV_NOOP at 0x0000000cdc34c360
Crashed Thread: 7

Application Specific Information:
Exception 1, Code 1, Subcode 55234052960 >
KERN_INVALID_ADDRESS at 0xcdc34c360.

Thread 7 Crashed:
0   libobjc.A.dylib                 0x312ccec14         objc_retain
1   CFNetwork                       0x3248c4168         _CFStreamErrorFromCFError
2   CFNetwork                       0x3248c4630         _CFStreamErrorFromCFError
3   CFNetwork                       0x3249f78d0         _CFHTTPServerResponseEnqueue
4   CFNetwork                       0x3248a6954         CFURLConnectionInvalidateConnectionCache
5   CFNetwork                       0x3249d3ef4         CFHTTPCookieStorageUnscheduleFromRunLoop
6   Foundation                      0x3206ad000         URLRequest._unconditionallyBridgeFromObjectiveC
7   <AppName>                         0x102dea7f8         NetworkInstrumentationFeature.intercept (NetworkInstrumentationFeature.swift:155)
8   <AppName>                         0x102deadbc         thunk for closure
9   libdispatch.dylib               0x3326206a4         _dispatch_call_block_and_release
10  libdispatch.dylib               0x3326222fc         _dispatch_client_callout
11  libdispatch.dylib               0x332629890         _dispatch_lane_serial_drain
12  libdispatch.dylib               0x33262a3c0         _dispatch_lane_invoke
13  libdispatch.dylib               0x332635000         _dispatch_root_queue_drain_deferred_wlh
14  libdispatch.dylib               0x332634874         _dispatch_workloop_worker_thread
15  libsystem_pthread.dylib         0x3f6671960         _pthread_wqthread
```

```
OS Version: iOS 17.2.1 (21C66)
Report Version: 104

Exception Type: EXC_CRASH (SIGABRT)
Crashed Thread: 16

Application Specific Information:
-[Swift.__StringStorage count]: unrecognized selector sent to instance 0x2802cdf80

Thread 16 Crashed:
0   CoreFoundation                  0x32f1c469c         __exceptionPreprocess
1   libobjc.A.dylib                 0x31f7efc7c         objc_exception_throw
2   CoreFoundation                  0x32f254fd8         -[NSObject(NSObject) doesNotRecognizeSelector:]
3   CoreFoundation                  0x32f109e04         ___forwarding___
4   CoreFoundation                  0x32f24a94c         __forwarding_prep_0___
5   CFNetwork                       0x33148f2b8         _CFHTTPServerResponseEnqueue
6   CFNetwork                       0x331476be0         CFHTTPCookieStorageUnscheduleFromRunLoop
7   Foundation                      0x32d436718         URLRequest.allHTTPHeaderFields.getter
8   <AppName>                         0x1072741c4         NetworkInstrumentationFeature.extractTrace (NetworkInstrumentationFeature.swift:246)
9   <AppName>                         0x107272ba4         [inlined] NetworkInstrumentationFeature.extractTrace
10  <AppName>                         0x107272ba4         NetworkInstrumentationFeature.intercept (NetworkInstrumentationFeature.swift:169)
11  <AppName>                         0x107272dbc         thunk for closure
12  libdispatch.dylib               0x33f0c06a4         _dispatch_call_block_and_release
13  libdispatch.dylib               0x33f0c22fc         _dispatch_client_callout
14  libdispatch.dylib               0x33f0c9890         _dispatch_lane_serial_drain
15  libdispatch.dylib               0x33f0ca3c0         _dispatch_lane_invoke
16  libdispatch.dylib               0x33f0d5000         _dispatch_root_queue_drain_deferred_wlh
17  libdispatch.dylib               0x33f0d4874         _dispatch_workloop_worker_thread
18  libsystem_pthread.dylib         0x403119960         _pthread_wqthread
19  libsystem_pthread.dylib         0x403119a00         start_wqthread
```

```
OS Version: iOS 17.2.1 (21C66)
Report Version: 104

Exception Type: EXC_CRASH (SIGABRT)
Crashed Thread: 8

Application Specific Information:
-[__NSMallocBlock__ objectForKey:]: unrecognized selector sent to instance 0x280c16980

Thread 8 Crashed:
0   CoreFoundation                  0x33ce1469c         __exceptionPreprocess
1   libobjc.A.dylib                 0x32d42fc7c         objc_exception_throw
2   CoreFoundation                  0x33cea4fd8         -[NSObject(NSObject) doesNotRecognizeSelector:]
3   CoreFoundation                  0x33cd59e04         ___forwarding___
4   CoreFoundation                  0x33ce9a94c         __forwarding_prep_0___
5   CFNetwork                       0x33f0df8b8         _CFHTTPServerResponseEnqueue
6   CFNetwork                       0x33efb48c8         _CFStreamErrorFromCFError
7   CFNetwork                       0x33f0c6b30         CFHTTPCookieStorageUnscheduleFromRunLoop
8   Foundation                      0x33b0875b0         URLRequest.value
9   <AppName>                         0x1052a6c0c         NetworkInstrumentationFeature.intercept (NetworkInstrumentationFeature.swift:173)
10  <AppName>                         0x1052a6dbc         thunk for closure
11  libdispatch.dylib               0x34cd106a4         _dispatch_call_block_and_release
12  libdispatch.dylib               0x34cd122fc         _dispatch_client_callout
13  libdispatch.dylib               0x34cd19890         _dispatch_lane_serial_drain
14  libdispatch.dylib               0x34cd1a3c0         _dispatch_lane_invoke
15  libdispatch.dylib               0x34cd25000         _dispatch_root_queue_drain_deferred_wlh
16  libdispatch.dylib               0x34cd24874         _dispatch_workloop_worker_thread
17  libsystem_pthread.dylib         0x410d69960         _pthread_wqthread
18  libsystem_pthread.dylib         0x410d69a00         start_wqthread
```

</p>
</details> 

### Reproduction

This is the code we're using to set this up. Like I said, we can't reproduce the crashes locally:

```swift
let networkHosts: Set<String> = ["some-domain.com"]
RUM.enable(with: RUM.Configuration(
    applicationID: "<app-id>",
    sessionSampleRate: 100,
    urlSessionTracking: RUM.Configuration.URLSessionTracking(
        firstPartyHostsTracing: .trace(hosts: networkHosts))))

Trace.enable(
        with: Trace.Configuration(
          urlSessionTracking: .init(firstPartyHostsTracing: .trace(
              hosts: networkHosts,
              sampleRate: 100)),
          networkInfoEnabled: true))

 URLSessionInstrumentation.enable(
      with: .init(
        delegateClass: URLTaskDelegate.self))

```

Then this is in essence how we make network requests:

```swift  
static let sessionDelegate = SessionDelegate()
static let session =  URLSession(
      configuration: URLSessionConfiguration.default,
      delegate: sessionDelegate,
      delegateQueue: nil)

let taskDelegate = URLTaskDelegate()
return try await session.data(for: request, delegate: taskDelegate)
```

---

#### Datadog SDK versions:

`2.6.0`

#### Last stable Datadog SDK version:

`2.6.0`

#### Volume:

~1%

#### OS version:

iOS 16.1 through iOS 17.3

#### Deployment Target:

iOS 16.1

#### Device version:

All kinds of devices: iPhone SE, XS through iPhone 15, and all sorts of iPads.

#### Environment:

No. It seems to be evenly distributed across users.

**Comment by jpsim** [comment]
We're seeing the same thing in our app starting with 2.6.0. Our usage of DataDog and URLSession is similar to what Javi mentioned.

**Comment by maxep** [comment]
Hi @JaviSoto 👋  Thank you for the detailed report, it's very helpful.

We are investigating and we will keep you posted!

**Comment by JackTudor96** [comment]
We're also experiencing this issue in our app

**Comment by ganeshnj** [comment]
I have a hunch that this is because of synchronization problem but I have not been able to reproduce the issue yet.

Even with 1000s of request made in parallel given `RUM`, `Trace` and `URLSessionInstrumentation` is enabled

For example
```swift
await withTaskGroup(of: Void.self) { group in
    for _ in 0..<1000 {
        group.addTask {
            let (data, response) = try! await session.data(for: request, delegate: taskDelegate)
            let httpResponse = response as! HTTPURLResponse
            print(httpResponse.statusCode)
            // json response
            let json = try! JSONSerialization.jsonObject(with: data, options: [])
            print(json)
        }
    }
}
```

Stay tuned.

**Comment by JaviSoto** [comment]
Hey! I also spent some time trying to reproduce it in a fresh project and have so far been unable to.

Is there anything in the 2.6.0 diff that may point at a possible cause?

**Comment by ganeshnj** [comment]
https://github.com/DataDog/dd-sdk-ios/releases/tag/2.6.0

We had one small [bug fix](https://github.com/DataDog/dd-sdk-ios/pull/1609) which was related to network instrumentation.

**Comment by JaviSoto** [comment]
Ah yeah I think that's likely the bug: `URLSessionTask` isn't thread-safe, and we're reading its `currentRequest`, which is a mutable property. If another thread happens to be modifying that property at the same time, that code may read a "partially written" pointer, which would explain the crashes we see, where trying to call methods on that `URLRequest` we end up reading garbage memory and crashing.

The solution may be to read the `task.currentRequest` synchronously from `func intercept(task:…` instead of waiting until we're in the private `queue`, since that method will be called from the same thread that the URL system is calling us from

**Comment by JaviSoto** [comment]
`URLSessionTask.currentRequest`'s documentation says:

> As the connection performs the load, the request may change as a result of protocol canonicalization or due to following redirects. This property provides the current value of the request.

So maybe one way to repro this is with similar code to the repro code you showed us, but making sure that the requests you're making require redirections?

**Comment by ganeshnj** [comment]
Can you share snippet how are you accessing the `currentRequest` (just the setup), which will help us understanding the use case. Given `URLSessionTask` is not part of the SDK, there is no easy way to make `currentRequest` thread safe.

Off the top of my head, I can see two options
- Either provide access to the queue outside the SDK (or some injection)
- Do some swizzling on the getter of `currentRequest`

**Comment by JaviSoto** [comment]
Oh, we're not mutating that `currentRequest` directly, but the URL Session system will when the request hits a redirection for example. I'm hypothesizing that that's what may be happening. After sending that message I did some more research and realized that `URLSessionTask` is actually marked as `@unchecked Sendable`, so that code in Datadog _should_ be correct. However, it's still worth trying out my suggested fix:

```diff
 func intercept(task: URLSessionTask, additionalFirstPartyHosts: FirstPartyHosts?) {
     // Get the current trace context from all handlers.
     let traceContexts = handlers.compactMap { $0.traceContext() }

-    queue.async { [weak self] in
-        guard let self = self, let request = task.currentRequest else {
+    guard let request = task.currentRequest else { return }
+
+    queue.async { [weak self] in
+      guard let self = self else {
             return
         }
```

**Comment by ganeshnj** [comment]
I have created a high priority issue internally to address this. Thanks for the idea, we will make sure that rather than holding the `currentRequest`, we can only target specific required properties (immutable ie copying).

**Comment by JaviSoto** [comment]
Hi! Was wondering if you had any updates for us on this 🙏  we'd love to be able to turn on RUM again. Thank you!

**Comment by djtarazona** [comment]
Sorry to bump - I'm curious if there is an update as well. We updated to 2.8.1 for the privacy manifest updates and have been seeing this crash since then.

**Comment by trevor-holliday-instacart** [comment]
@ganeshnj bumping this as well. Is there any update on resolving this crash?

**Comment by maxep** [comment]
Hello 👋 

Sorry for the late reply, no update yet to share but it's top priority!

**Comment by ganeshnj** [comment]
Apologies @djtarazona @trevor-holliday-instacart. I was away. 

Thanks @maxep for the update on the status.

**Comment by maxep** [comment]
Hey everyone 👋 

The [2.9.0](https://github.com/DataDog/dd-sdk-ios/releases/tag/2.9.0) is out and include a fix.

Be aware that we are not able to reproduce the crash, we have based the fix on the information you have provided and observation we made, see more info in #1767. Please re-open this ticket if you are still seeing the issue after the update.

Thanks again for your help!

**Comment by djtarazona** [comment]
@maxep I think I just got this crash (or related) while I was running with the debugger attached:

<img width="1665" alt="Screenshot 2024-04-15 at 22 08 52" src="https://github.com/DataDog/dd-sdk-ios/assets/337986/4b1a1375-6edd-4c40-8eb0-e21420b4a50a">

**Comment by Limwin94** [comment]
@maxep 
I am currently experiencing the same issue as @djtarazona

**Comment by Limwin94** [comment]
There's a new issue with this.
I am currently experiencing the same issue(allHTTPHeaderFields Bad Access) in 2.9.0

**Comment by ganeshnj** [comment]
Thinking aloud here, assuming this is a concurrency issue, while we do swizzling for other methods for instrumentation, we can swizzle the `allHTTPHeaderFields` getter and make it thread safe. That's how we can make sure it stays thread safe in Datadog SDK and other places.

**Comment by joseewu** [comment]
👋 we are in 2.10.0 and encounter the same issue as well.

**Comment by ncreated** [comment]
The fix was merged. It will be available in next release.

**Comment by djtarazona** [comment]
> @djtarazona By any chance, are you able to reproduce this and dump the actual request.allHTTPHeaderFields?

Sorry @ncreated, I've been quite busy and haven't gotten to this. Due to the crash, we downgraded the version of `dd-sdk-ios` we were using, so I wasn't running into the crash anymore.

Now that a potential fix is coming out in the next release, we'll give it a try again. Thanks!

**Comment by JaviSoto** [comment]
Hey! Thank you for the fix, that looks certainly promising!!

Any chance we can get a release soon with the fix? 🙏

**Comment by ncreated** [comment]
Hey @JaviSoto, [2.12.0](https://github.com/DataDog/dd-sdk-ios/releases/tag/2.12.0) is out and it should fix this problem. Please let us know if the issue persists by any chance.

**Comment by JaviSoto** [comment]
Sweet thank you very much! We have re-enabled this and will let you know if we run into any more issues 🙏

**Comment by JackTudor96** [comment]
Hi there, we're using 2.12.0 and I think we're encountering this issue still. @ncreated are you able to advise at all?

<img width="1410" alt="Screenshot 2024-06-10 at 16 41 22" src="https://github.com/DataDog/dd-sdk-ios/assets/32430158/fc24668b-5b4f-4f21-857f-2e99312d778f">

**Comment by ncreated** [comment]
Thanks @JackTudor96 . I wonder if this is now only limited to Alamofire and `DatadogAlamofireExtension`. As the last attempt on solving this problem, we got rid of `allHTTPHeaderFields` usage in our code - in all scenarios, except Alamofire instrumentation which still [uses it in `URLSessionInterceptor`](https://github.com/DataDog/dd-sdk-ios/blob/46c68160b13bb1cd33b5af78959b749c89c20bff/DatadogInternal/Sources/NetworkInstrumentation/URLSession/URLSessionInterceptor.swift#L101).

It would be very helpful if one can confirm the issue is gone / not gone in non-Alamofire setups. @JaviSoto @djtarazona @trevor-holliday-instacart @joseewu does the issue still persist in `2.12` on your side? Do you use Alamofire?

**Comment by JaviSoto** [comment]
We still haven't rolled out an update with this new version so I can't yet say whether the crash is back. However, I can say that we don't use Alamofire, so at least the original crash wasn't related to that.

**Comment by JackTudor96** [comment]
@ncreated I'm going to roll out another release with some additional logging on our end based on your feedback, once I've got some of the results is there an email I can reach you on?

**Comment by ncreated** [comment]
@JackTudor96 Please do 🙏. You can find me on [Datadog's public Slack](https://chat.datadoghq.com/), under my nickname (`ncreated`).

**Comment by djtarazona** [comment]
@ncreated We've updated to `2.13` and are getting the same crash as @JackTudor96.

Did you guys find a solution? Note we are using Alamofire.

<img width="1667" alt="Screenshot 2024-06-25 at 13 41 29" src="https://github.com/DataDog/dd-sdk-ios/assets/337986/a2875aff-03b9-4e97-9670-43cc9b016020">

**Comment by djtarazona** [comment]
Great, thanks! Really appreciate the tenacity on this. 🙏

Giving `URLSessionInstrumentation.enable(with:)` a shot. We won't be pushing this out to our Production users until we have more confidence this is stable. With our team's internal usage, I should have something to report back in about a week.

**Comment by djtarazona** [comment]
> Giving `URLSessionInstrumentation.enable(with:)` a shot. We won't be pushing this out to our Production users until we have more confidence this is stable. With our team's internal usage, I should have something to report back in a week.

We've had no crashes on `2.13.0` using `URLSessionInstrumentation.enable(with:)` instead of `DatadogAlamofireExtension`. 🎉

We've had this out to our Production users for over a week - looks like this solution will stick. Thanks a bunch!

**Comment by JaviSoto** [comment]
I can also confirm we've seen no crashes with this enabled since we updated to the last version. Thanks everyone!

**Comment by ncreated** [comment]
@djtarazona @JaviSoto thanks a ton for confirming this worked. It was a long battle, glad we managed to fix the problem 💪
